source/agora/node/EnrollmentManager.d

```diff
@@ -0,0 +1,416 @@
+        this.data = new EnrollmentData();
+
+        // public key
+        this.data.K = this.node_keypair.address;
+
+        // cycle length
+        this.data.n = 1008; // freezing period / 2
+
+        // generate random seed value and save
+        this.random_seed = Scalar.random();
+        auto seed_file = File(this.random_seed_path, "w");
+        this.random_seed.serialize((scope bytes) => seed_file.rawWrite(bytes));
+
+        // nth image of random seed
+        this.data.X = hashFull(this.random_seed);
+        foreach (i; 0 .. this.data.n-1)
+            this.data.X = hashFull(this.data.X);
+
+        // signature noise
+        this.schnorr_pair = Pair.random();
+        this.data.R = this.schnorr_pair.V;
+
+        // signature
+        Hash hash = hashFull(*this.data);
+        data.S = signHash(this.node_keypair, schnorr_pair, hash);
+
+        return *this.data;
+    }
+
+    /// Ditto, only used in unittests
+    version (unittest)
+    public static EnrollmentData createEnrollmentData (KeyPair key_pair) @trusted
+    {
+        auto data = new EnrollmentData();
+
+        // public key
+        data.K = key_pair.address;
+
+        // cycle length
+        data.n = 1008; // freezing period / 2
+
+        // nth image of random seed
+        data.X = Scalar.random().hashFull();
+        foreach (i; 0 .. data.n-1)
+            data.X = hashFull(data.X);
+
+        // signature noise
+        auto schnorr_pair = Pair.random();
+        data.R = schnorr_pair.V;
+
+        // signature
+        Hash hash = hashFull(*data);
+        data.S = signHash(key_pair, schnorr_pair, hash);
+
+        return *data;
+    }
+
+    // sign enrollment data with schnorr key pairs
+    private static Signature signHash (ref KeyPair key_pair, 
+        ref Pair schnorr_pair, ref Hash hash)
+    {
+        auto pk = Scalar(key_pair.seed[]);
+        auto pk_point = pk.toPoint();
+        auto signature = sign(pk, pk_point, schnorr_pair.V, schnorr_pair.v,
+            hash[]);
+
+        return signature;
+    }
+
+    /***************************************************************************
+
+        Check if the EnrollmentData is right
+
+    ***************************************************************************/
+
+    public bool acceptEnrollment (const ref EnrollmentData enroll) @safe
+    {
+        if (!this.add(enroll))
+        {
+            log.info("Rejected enroll: {}", enroll);
+            return false;
+        }
+
+        return true;
+    }
+
+    /***************************************************************************
+
+        Check if a enrollment data hash exists in the transaction pool.
+
+        Params:
+            enroll = the enrollment data hash
+
+        Returns:
+            true if the enrollment manager has the enrollment data hash
+
+    ***************************************************************************/
+
+    public bool hasEnrollmentDataHash (const ref EnrollmentData enroll) @trusted
```

Comment:
This should be renamed to `hasEnrollmentData`. From the name it seemed like the user would provide the hash to the function, not the data.